### PR TITLE
report errors on where clauses in optional match queries.

### PR DESCRIPTION
**Pull request by xtcyclist**
## What type of PR is this?
- [X] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
Close https://github.com/vesoft-inc/nebula/issues/5272

#### Description:
To support queries like `OPTIONAL MATCH <pattern> WHERE <condition>` has been a beta feature, in which the `<condition>` cannot be evaluated correctly in all cases in the current version. A simpler query like `OPTIONAL MATCH <pattern>` without a where clause, on the other hand, can be evaluated correctly.

## How do you solve it?
Report errors on the where clauses in optional match queries.

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [x] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [x] Documentation affected (Please add the label if documentation needs to be modified.)
- [x] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [x] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

**If users have written queries like `optional match ... where...`, it may have appeared to be running ok, although it is not. With this PR, such queries would run into error messages.**

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/parser/parser.yy [hunk 1]
```diff
@@ -1682,7 +1682,11 @@ match_clause
         $$ = new MatchClause($2, $3, false/*optional*/);
     }
     | KW_OPTIONAL KW_MATCH match_path_list where_clause {
-        $$ = new MatchClause($3, $4, true);
+        if ($4 != nullptr) {
+            throw nebula::GraphParser::syntax_error(@4, "Where clause in optional match is not supported.");
```
Comment (yixinglu) [inline src/parser/parser.yy:1686]: delete $3 before throwing error to avoid memory leak of $3.

Comment (yixinglu) [reply src/parser/parser.yy:1686]: and delete $4

Comment (xtcyclist) [reply src/parser/parser.yy:1687]: Fixed.

**Review by yixinglu — APPROVED** [review]
LGTM

**Review by codesigner — APPROVED** [review]

**Comment by abby-cyber** [comment]
https://github.com/vesoft-inc/nebula-docs-cn/pull/2515

**Pull request merged**